rfc3164/machine.go.rl

```diff
@@ -0,0 +1,204 @@
+    errTimestamp      = "expecting a Stamp timestamp [col %d]"
+    errHostname       = "expecting an hostname (from 1 to max 255 US-ASCII characters) [col %d]"
+    errTag            = "expecting an alphanumeric tag (max 32 characters) [col %d]"
+    errContentStart   = "expecting a content part starting with a non-alphanumeric character [col %d]"
+    errContent        = "expecting a content part composed by visible characters only [col %d]"
+    errParse          = "parsing error [col %d]"
+)
+
+%%{
+machine rfc3164;
+
+include common "common.rl";
+
+# unsigned alphabet
+alphtype uint8;
+
+action mark {
+    m.pb = m.p
+}
+
+action set_prival {
+    output.priority = uint8(common.UnsafeUTF8DecimalCodePointsToInt(m.text()))
+	output.prioritySet = true
+}
+
+action set_timestamp {
+    if t, e := time.Parse(time.Stamp, string(m.text())); e != nil {
+        m.err = fmt.Errorf("%s [col %d]", e, m.p)
+        fhold;
+        fgoto fail;
+    } else {
+        output.timestamp = t
+        output.timestampSet = true
+    }
+}
+
+action set_hostname {
+    output.hostname = string(m.text())
+}
+
+action set_tag {
+    output.tag = string(m.text())
+}
+
+action set_content {
+    output.content = string(m.text())
+}
+
+action set_message {
+    output.message = string(m.text())
+}
+
+action err_prival {
+    m.err = fmt.Errorf(errPrival, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_pri {
+    m.err = fmt.Errorf(errPri, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_timestamp {
+    m.err = fmt.Errorf(errTimestamp, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_hostname {
+	m.err = fmt.Errorf(errHostname, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_tag {
+    m.err = fmt.Errorf(errTag, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_contentstart {
+    m.err = fmt.Errorf(errContentStart, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_content {
+    m.err = fmt.Errorf(errContent, m.p)
+	fhold;
+    fgoto fail;
+}
+
+pri = ('<' prival >mark %from(set_prival) $err(err_prival) '>') @err(err_pri);
+
+timestamp = (datemmm sp datemday sp hhmmss) >mark %set_timestamp @err(err_timestamp);
+
+# (todo) > RFC3164 says something about its maximum length?
+hostname = hostnamerange >mark %set_hostname $err(err_hostname);
+
+# Section 4.1.3
+tag = (print - [ :\[]){1,32} >mark %set_tag @err(err_tag); # note(leodido) > alnum{1,32} is too restrictive (no dashes, see https://tools.ietf.org/html/rfc2234#section-2.1)
+
+visible = print | 0x80..0xFF;
+
+# The first not alphanumeric character starts the content part of the message part
+contentval = !alnum @err(err_contentstart) >mark print* %set_content @err(err_content);
+
+
+content = '[' contentval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
+
+mex = visible+ >mark %set_message;
+
+msg = (tag content? ':' sp)? mex;
+
+fail := (any - [\n\r])* @err{ fgoto main; };
+
+main := pri timestamp sp hostname sp msg;
```

Comment:
Do you imagine in the future something like this?

```
(pri (timestamp (sp hostname sp)?)?)? msg;
```

Reply:
I agree that this PR does not need to handle 4.3.2 (valid PRI + no timestamp/invalid timestamp) and 4.3.3 (no or unidentifiable PRI).

Nevertheless, I think this parser should catch these situations - eg., throwing (non-generic) errors.
So that later, a user implementing a relay, or simply a wrapper around this parser, could detect those errors (for example no PRI, 4.3.3) and behave accordingly.

Clearly, this is easily doable as long I implement best-effort mode for this parser too. In this way when errors occur we can still recover the good parts of the message (useful in 4.3.2 use case for example).

Reply:
Smart.  I like that point.  Use errors.